h/api/schemas.py

```diff
@@ -90,9 +205,75 @@ class CreateAnnotationSchema(object):
 
     def __init__(self, request):
         self.request = request
-        self.structure = AnnotationSchema()
+        if self.request.feature('postgres_write'):
+            self.structure = AnnotationSchema()
+        else:
+            self.structure = LegacyAnnotationSchema()
 
     def validate(self, data):
+        if not self.request.feature('postgres_write'):
+            return self._legacy_validate(data)
+
+        appstruct = self.structure.validate(data)
+
+        new_appstruct = {}
+
+        # Some fields are not to be set by the user, ignore them.
+        for field in PROTECTED_FIELDS:
+            appstruct.pop(field, None)
+
+        new_appstruct['userid'] = self.request.authenticated_userid
+
+        new_appstruct['target_uri'] = appstruct.pop('uri', '')
+        new_appstruct['text'] = appstruct.pop('text', '')
+        new_appstruct['tags'] = appstruct.pop('tags', [])
+
+        # Replace the client's complex permissions object with a simple shared
+        # boolean.
+        if appstruct.pop('permissions')['read'] == [new_appstruct['userid']]:
+            new_appstruct['shared'] = False
+        else:
+            new_appstruct['shared'] = True
+
+        # The 'target' dict that the client sends is replaced with a single
+        # annotation.target_selectors whose value is the first selector in
+        # the client'ss target.selectors list.
+        # Anything else in the target dict, and any selectors after the first,
+        # are discarded.
+        target = appstruct.pop('target', [])
+        if target:  # Replies and page notes don't have 'target'.
+            target = target[0]  # Multiple targets are ignored.
+            new_appstruct['target_selectors'] = target['selector']
+
+        new_appstruct['groupid'] = appstruct.pop('group', '__world__')
+
+        new_appstruct['references'] = appstruct.pop('references', [])
+
+        # Replies always get the same groupid as their parent. The parent's
+        # groupid is added to the reply annotation later by the storage code.
+        # Here we just delete any group sent by the client from replies.
+        if new_appstruct['references'] and 'groupid' in new_appstruct:
+            del new_appstruct['groupid']
+
+        new_appstruct['extra'] = appstruct
+
+        # Transform the "document" dict that the client posts into a convenient
+        # format for creating DocumentURI and DocumentMeta objects later.
+        document_data = appstruct.pop('document', {})
```

Comment:
Would it make sense to pass separate copies of `document_data`into the next two function calls? So we don't rely on these two functions not mutating this dictionary.